FlatBuffers for IPC

RobertBColton opened the pull request:
Ok, so we hit some limitations in #1115 that demonstrated perhaps Protocol buffers is not the best choice for ENIGMA. I did some shopping around and found FlatBuffers instead. So far I really like it, it was even simpler to set up, produces less boilerplate, and is actually geared towards video games.

https://google.github.io/flatbuffers/

I also used snake_case for the fields this time but I am confused about when to use table or struct. It would not let me use struct for Image because it has a byte array for the image data (apparently FlatBuffers only allows scalar values in structs). I would appreciate some general rule of thumbs for doing this and also other feedback on the implementation.

Also contributing the codegen again, because we said we want to make it easy for people to build.

Edit: I discovered this nugget, which is supportive of snake_case:
> Table and struct field names: snake_case. This is translated to lowerCamelCase automatically for some languages, e.g. Java.

JoshDreamland commented on CompilerSource/backendflat/Background.fbs:
```diff
@@ -0,0 +1,27 @@
+table Image {
+	width:int;
+	height:int;
+	data:[byte];
```
Format? If MIME, comment it; if RGBA, use int; if any of these, use a separate message and a `oneof`.

RobertBColton replied:
I'm confused, are you asking me to add the MIME as a field? I believe we were just sending the raw image data prior, so it's always in the same format. This means the IDE, and not the compiler, requires an image loader. Please elaborate and clarify further.

JoshDreamland commented on CompilerSource/backendflat/Background_generated.h:
```diff
@@ -0,0 +1,295 @@
+// automatically generated by the FlatBuffers compiler, do not modify
```
It amazes me that a Google project would write out such poor grammar... I sort of wish this was your doing....

RobertBColton replied:
I don't understand? Where's the mistake? All I see is that it's technically not *automatic* because I had to invoke the compiler.

RobertBColton replied:
Or do you mean the rest of this file?

RobertBColton commented:
I'll answer your first questions first. So the flatc.exe I downloaded for Windows is 896kb's compared to protoc.exe's 4,315kb's.

FlatBuffers seems to involve a lot fewer steps in building it from source:
https://github.com/google/protobuf/blob/master/src/README.md
http://google.github.io/flatbuffers/flatbuffers_guide_building.html

These are the sizes of a ZIP of each project's master branch:
* FlatBuffers: 776 KB
* Protobuf: 6.28 MB

So to answer your question, it's a lot smaller, and as a result it's actually feasible for us to include it in our source chain.

RobertBColton commented:
I have some lingering questions myself I'd like you to answer:

Quoted from this page:
https://google.github.io/flatbuffers/flatbuffers_guide_writing_schema.html

* Do you want to manually assign id's?
> Older code will simply ignore the new field. If you want to have flexibility to use any order for fields in your schema, you can manually assign ids
* Do you want to use fixed width types for most fields? i.e. ``int32`` instead of plain ``int``
> The type names in parentheses are alias names
* Do you want to add file identifiers?
> You can specify in a schema, similar to root_type, that you intend for this type of FlatBuffer to be used as a file format:

RobertBColton commented:
Regarding the directory, I named it "backendflat" because the idea was to keep the folders close together so it's obvious what's in there. Especially because we want to use this to replace the entire backend we have now, eventually deleting the old one when the time comes. But in the interest of time, I've renamed it "ipc" per your request.

I also moved the codegen to a subfolder called "codegen" which is more appropriate because there is a separate concept of binaries in FlatBuffers so "/bin" would have been confusing.

I generated the codegen using these commands:
```bash
"C:\Users\Owner\Desktop\flatc_windows_exe\flatc.exe" -o ./codegen/ --cpp Image.fbs
"C:\Users\Owner\Desktop\flatc_windows_exe\flatc.exe" -o ./codegen/ --cpp Background.fbs
```

RobertBColton commented:
I have some lingering questions myself I'd like you to answer:

Quoted from this page:
https://google.github.io/flatbuffers/flatbuffers_guide_writing_schema.html

* Do you want to manually assign id's?
> Older code will simply ignore the new field. If you want to have flexibility to use any order for fields in your schema, you can manually assign ids

> The solution of course is that you should not be generating binary data before your schema changes have been committed, ensuring consistency with the rest of the world. If this is not practical for you, use explicit field ids, which should always generate a merge conflict if two people try to allocate the same id.
* Do you want to use fixed width types for most fields? i.e. ``int32`` instead of plain ``int``
> The type names in parentheses are alias names

> FlatBuffers supports the full range of integer sizes, **_so try to pick the smallest size needed_**, rather than defaulting to int/long.
* Do you want to add file identifiers?
> You can specify in a schema, similar to root_type, that you intend for this type of FlatBuffer to be used as a file format: 
* Do you want to assign default values? I think we do because then we have them in a single place, and the IDE won't have to redundantly redefine the defaults.

JoshDreamland commented on CompilerSource/ipc/codegen/Background_generated.h:
```diff
@@ -0,0 +1,222 @@
+#ifndef FLATBUFFERS_GENERATED_BACKGROUND_H_
+#define FLATBUFFERS_GENERATED_BACKGROUND_H_
+
+#include "flatbuffers/flatbuffers.h"
```
Where is this header coming from? We don't ship it on Windows. Is it in the AUR? The Debian repos?

RobertBColton replied:
I haven't figured that out yet, I guess that's supposed to be the only file you need for the runtime or w/e.

RobertBColton replied:
I'll just copy the comment from the plugin https://github.com/enigma-dev/lgmplugin/blob/796232a0b961db2fdc979f5a64ccdd6f2acbe95c/org/enigma/backend/util/Image.java#L21

BGRA is what GameMaker still is too, even in Studio.

RobertBColton replied:
It does look like that's what it is from:
https://google.github.io/flatbuffers/

> Use flatc (the FlatBuffer compiler) to generate a C++ header (or Java/C#/Go/Python.. classes) with helper classes to access and construct serialized data. This header (say mydata_generated.h) only depends on flatbuffers.h, which defines the core functionality.

So we can do that, just wait until we get over to writing the EGM reading in C++ first.

RobertBColton replied:
It does look like that's what it's from:
https://google.github.io/flatbuffers/

> Use flatc (the FlatBuffer compiler) to generate a C++ header (or Java/C#/Go/Python.. classes) with helper classes to access and construct serialized data. This header (say mydata_generated.h) only depends on flatbuffers.h, which defines the core functionality.

So we can do that, just wait until we get over to writing the EGM reading in C++ first.

JoshDreamland replied:
That's great. I don't have a problem with checking it in in this PR, if you like. I also don't mind waiting (but why would you check in the codegen and not its dependencies?).

RobertBColton commented:
Ok, I can sum up your fears real fast by saying **_you're way off_**.

They don't shut up about JSON in the docs:
https://google.github.io/flatbuffers/flatbuffers_guide_tutorial.html
> JSON with FlatBuffers

You can generate a full object API for reflection:
https://google.github.io/flatbuffers/flatbuffers_guide_tutorial.html
> If this is not sufficient, other ways of mutating FlatBuffers may be supported in your language through an object based API (`--gen-object-api`) or reflection. See the individual language documents for support.

https://google.github.io/flatbuffers/flatbuffers_guide_using_schema_compiler.html
> * `--reflect-types` : Add minimal type reflection to code generation.
> * `--reflect-names` : Add minimal type/name reflection.
> * `--gen-object-api` : Generate an additional object-based API. This API is more convenient for object construction and mutation than the base API, at the cost of efficiency (object allocation). Recommended only to be used if other options are insufficient.

RobertBColton commented:
> Do FlatBuffers expose a way to check if a buffer actually contains a given field?
Yes:
https://google.github.io/flatbuffers/flatbuffers_guide_writing_schema.html
> Testing whether a field is present in a table

JoshDreamland commented:
Excellent. I'm still minorly worried about type resizing, but we'll cross that bridge when we come to it. Let's go ahead.

I have verified that FlatBuffer is available in the AUR, and I expect it to be in some Debian repository as well, if not in Ubuntu Universe.

RobertBColton commented on CompilerSource/ipc/Image.fbs:
```diff
@@ -0,0 +1,8 @@
+
```
Nah, this one is just kind of my own quirk, I don't ever like writing code on the first line of any source in any language unless it's a documentation comment, an import, or an include, etc. I try to keep my code where I can hit it easily with my mouse, which means away from the edges of my computer screen. This is not the result of any faulty hardware I am using or anything, just a personal preference to keep code away from the edges of the file. I'm open to feedback about this though haha.

RobertBColton commented on CompilerSource/ipc/Image.fbs:
```diff
@@ -0,0 +1,8 @@
+table Image {
+  width:int;
+  height:int;
+  data:[byte];
```
I can do that, FlatBuffers supports c-style comments as well as `/// comments` which get included/copied in the codegen. I didn't think about commenting it, sorry.

RobertBColton replied:
I'll just copy the comment from the plugin https://github.com/enigma-dev/lgmplugin/blob/796232a0b961db2fdc979f5a64ccdd6f2acbe95c/org/enigma/backend/util/Image.java#L21

BGRA is what GameMaker still is too, even in Studio.

RobertBColton commented:
Ok, I had to look into the rest of the issues, and I have good news and like 2 and a half questions.

# Fixed-width Types
> You can't change types of fields once they're used, with the exception of same-size data where a reinterpret_cast would give you a desirable result, e.g. you could change a uint to an int if no values in current data use the high bit yet.

So should I go for int32 in most of these cases? I think for background texture stuff we are going to want a 32 bit integer, which is pretty standard for graphics. Again, if this is what we are going to do, I could use a good rule of thumb for this.

# Manual Ids
> Yes; my understanding is that this is crucial to the backward-compatibility aspect of this serialization layer.

> Whatever the circumstance, if (5) doesn't work, we have a problem.

It is **_not_** and yes that works, as long as it's at the end not the beginning. From:
https://google.github.io/flatbuffers/flatbuffers_guide_writing_schema.html

> Each field is optional: It does not have to appear in the wire representation, and you can choose to omit fields for each individual object. As a result, you have the flexibility to add fields without fear of bloating your data. This design is also FlatBuffer's mechanism for forward and backwards compatibility.

> You can add new fields in the schema **_ONLY at the end of a table_** definition. Older data will still read correctly, and give you the default value when read. Older code will simply ignore the new field. If you want to have flexibility to use any order for fields in your schema, you can manually assign ids (much like Protocol Buffers), see the id attribute below.

So you really only need manual ids if you want to change up the order and such. I suggest you to read that whole section titled "Tables" and answer the question do you want me to assign manual ids?

# Default Values
>  include it in the buffer descriptor

By "descriptor", do you mean like a field "attribute" as they are called in FlatBuffers?

> In flatbuffer, this happens anyway, so there's probably no harm (and no better way).

Actually, FlatBuffers does the opposite of protobuf, saving space. From:
https://google.github.io/flatbuffers/flatbuffers_guide_tutorial.html

> Note: Since we are passing 150 as the mana field, which happens to be the default value, the field will not actually be written to the buffer, since the default value will be returned on query anyway. This is a **_nice space savings_**, especially if default values are common in your data. It also means that you do not need to be worried of adding a lot of fields that are only used in a small number of instances, as it will not bloat the buffer if unused.

RobertBColton commented on CompilerSource/ipc/Background.fbs:
```diff
@@ -13,8 +13,8 @@ table Background {
-  horizontal_separation:int;
-  vertical_separation:int;
+  horizontal_spacing:int;
+  vertical_spacing:int;
```
I don't mean to nitpick, but I just like the field name being a little shorter. I think spacing is a lot clearer too, in addition to being more common terminology used in GUI frameworks.

RobertBColton commented:
# Default Values
I see what you are saying, and yes I do actually want the IDE to display the defaults that I am suggesting. For example, the default `horizontal_spacing` field for backgrounds is supposed to be `16`. I want the user to be able to see this, it kind of just immediately points out what the field is for and makes it more sensible. 

But I do see what you mean that the compiler could use a heuristic etc, to compute better defaults. I just can't think of a good example where we would want that. Another case is where objects should have the create, step, and draw events when first created. In some cases it may make more sense to consider default values based on the value of other fields.

You know... we could actually do it both ways, in the schema, and use heuristics just for certain fields where it's necessary.

# Manual Ids
> That said, we can probably add the field numbers **_when the time comes_** that we actually do want to insert a field somewhere other than the end of the table.

Let's do that for now to keep it simple. If there is a need to reorder them then that is probably involving a change that requires just rewriting the whole schema for said resource anyway.

# Fixed-width Types
> I assume the default int is int64, so yes, go ahead and use int32 for most things.

`int` is actually an alias for `int32` From:
https://google.github.io/flatbuffers/flatbuffers_guide_writing_schema.html

So should I just leave it like it is? Can we trust that they will always keep it as int32?

RobertBColton commented on CompilerSource/ipc/Path.fbs:
```diff
@@ -0,0 +1,23 @@
+
+  //backgroundRoomId:int; //This was used by LGM rather than ENIGMA; it doesn't need to be here.
+  snapX:int; //TODO: Find out if these are needed and decide on keeping them
+  snapY:int;
```
Shall we dump these now rather than later?

RobertBColton commented on CompilerSource/ipc/Path.fbs:
```diff
@@ -0,0 +1,23 @@
+  snapX:int; //TODO: Find out if these are needed and decide on keeping them
+  snapY:int;
+
+  points:[PathPoint];
```
This here is going to be a vector basically, is that ok? I can't find anywhere that FlatBuffers has a plain-old array field type. Because of this, we don't need to write the pointCount, should we anyway? Also consider that Image will have a vector of bytes. Is there any performance overhead involved here from not using plain arrays?

RobertBColton commented on CompilerSource/ipc/Path.fbs:
```diff
@@ -0,0 +1,23 @@
+
+table Path {
+  name:string;
+  id:int;
```
Ok, so now that I have added another resource type, this is an important question to ask. Because neither FlatBuffers nor protobuf have inheritance, what should we do about common fields like `name` and `id`?

* We can obviously just write them out for each resource, I don't care.
* We could make the resource types a union.
* We could use composition and store those fields in a struct referenced by all resources.

JoshDreamland replied:
I don't have a strong opinion on this; the two are related, sure, but that's basically all there is to identifying something. Maybe later we might want to add a creation timestamp? If we decided to do that, then having this separate message type would pay off. Otherwise, it's a bit pointless. Do whichever you feel looks best.

RobertBColton replied:
No, that's not what I'm getting at actually. I meant how Path and Background both have name, and for each new resource type I must type `name` and similarly for `id`.

RobertBColton commented on CompilerSource/ipc/Path.fbs:
```diff
@@ -0,0 +1,23 @@
+
+  //background_room_id:int; //This was used by LGM rather than ENIGMA; it doesn't need to be here.
+  snap_x:int; //TODO: Find out if these are needed and decide on keeping them
+  snap_y:int;
```
Shall we dump these now rather than later in light of our discoveries about manual ids?

JoshDreamland replied:
Definitely extract these into their own message. Name it something like "PathEditorMetadata." I think they're still useful for EGM/GMX serialization. Actually, does GMX store these? That should be signal, right there.

RobertBColton commented on CompilerSource/ipc/Path.fbs:
```diff
@@ -0,0 +1,23 @@
+  snap_x:int; //TODO: Find out if these are needed and decide on keeping them
+  snap_y:int;
+
+  points:[PathPoint];
```
This here is going to be a vector basically, is that ok? I can't find anywhere that FlatBuffers has a plain-old array field type. Because of this, we don't need to write the pointCount, should we anyway? Also consider that Image will have a vector of bytes. Is there any performance overhead involved here from not using plain arrays?

JoshDreamland replied:
A vector is great, and we should definitely *not* write the point count. The awesome thing about protobuf/flatbuf is that we don't have to worry about Java vector vs C++ vector vs JavaScript array—it just works. So let it just work.

I suspect the performance overhead can be eliminated using some kind of Reserve method, or at worst, AddAll. I don't know enough about FlatBuf to say for sure, though—proto doesn't have this problem (or rather, it always has this problem in the same way), because it always copies.

RobertBColton commented on CompilerSource/ipc/Path.fbs:
```diff
@@ -5,18 +5,22 @@ struct PathPoint {
+  background_room_id:int;
+  snap_x:int;
+  snap_y:int;
+}
```
Ok, I've extracted these out now since they are not used by the paths extension itself. Going back to the integer ids again, is it possible that if we did what me and @rpjohnst discussed last night and use names for serialization and ids at runtime, that we could eliminate the need to ever defrag ids? GameMaker had to do that from time to time so that the ids didn't get too large or also in the case that they got corrupted. I am thinking we could use names only in the serialization and then automatically assign ids upon deserialization and only use them internally. Is this possible? Can we achieve this?

RobertBColton commented:
Ok, I've added the Font resource schema now and realized I can compile this with fewer commands:
```bash
"C:\Users\Owner\Desktop\flatc_windows_exe\flatc.exe" -o ./codegen/ --cpp Image.fbs Background.fbs Path.fbs Font.fbs
```
How are we going to add this to our build system? Can I write a script here to just automatically find the (*.fbs) schema files? We could go with single-header output, since we will be serializing the tree structure, which itself must recursively include all of the other resource types.

Also, I could not use FontEditorMetadata like I did for the Path resource because bold, size, italic all have getter functions in GML and EDL.

RobertBColton commented on CompilerSource/ipc/Shader.fbs:
```diff
@@ -0,0 +1,15 @@
+
+enum ShaderType : byte { GLSL, GLSLES, HLSL9, HLSL11 }
```
Is this extensible? I know we are on Vulkan now, but what if we later want to add a new HLSL version in between HLSL and Vulkan at the end. This is an area where we will likely want to update somewhat frequently.

Enums are discussed here:
https://google.github.io/flatbuffers/flatbuffers_guide_writing_schema.html

RobertBColton commented on CompilerSource/ipc/Shader.fbs:
```diff
@@ -0,0 +1,15 @@
+  id:int;
+
+  type:ShaderType;
+  precompile:bool;
```
This precompile boolean is a field I added to ENIGMA that does not exist in GMS. Do we want to keep this around and continue going in this direction or remove it? I think it's useful to allow control over this.

RobertBColton commented on CompilerSource/ipc/Sound.fbs:
```diff
@@ -0,0 +1,27 @@
+
+enum SoundKind : byte { NORMAL, BACKGROUND_MUSIC, THREE_DIMENSIONAL, MULTIMEDIA_PLAYER }
```
I don't know if we want to start dropping some of this stuff from legacy GM that ENIGMA isn't using. This is one particular case, but we are also not including the new properties from GMS to specify the encoding and such. We also don't have code for that engine side. We had discussed changing this to offer some encoding options to make games smaller. There's also a "Type" in GMX which is completely distinct from "Kind" but both are serialized. Type has the options "Mono," "Stereo," or "3D." Thoughts?

RobertBColton commented on CompilerSource/ipc/GameInformation.fbs:
```diff
@@ -0,0 +1,18 @@
+  text:string;
+}
+
+root_type GameInformation;
```
Do we maybe want to consider dropping this resource altogether? I don't know how well it's used. GMS 2 has dropped it and replaced it with an instantiatable "Notes" resource which is plain text files. RTF again was a Windowsy thing. I don't use this feature anymore myself and wouldn't even use it for a quick and dirty game prototype. That said, if we drop it, I'd be happy to look into something like Markdown editing in the future since that's pretty standard now. We can also defer dropping this until later.

This thing also used to have a constructor with some useful defaults:
https://github.com/enigma-dev/enigma-dev/blob/a79faf94c27a7db0f17dc9e236c5116f08e7e738/CompilerSource/backend/resources/GameInformation.h#L30
I think that's the only resource in the original backend with defaults compiled into it too, which I found interesting. Not sure where to go with those.

RobertBColton commented on CompilerSource/ipc/Sprite.fbs:
```diff
@@ -0,0 +1,39 @@
+  smooth_edges:bool;
+  preload:bool;
+  origin_x:int;
+  origin_y:int;
```
Should I just use the Point struct here for origin? I don't know why we didn't do that in the original backend, but we can do that easily now `origin:Point;`.

RobertBColton commented on CompilerSource/ipc/Sprite.fbs:
```diff
@@ -0,0 +1,39 @@
+
+  bounding_box:SpriteBoundingBox;
+
+  subimages:[Image];
```
The original backend had SubImage as a simple wrapper around Image. I saw no reason for that, so I've simply used Image here. Is that ok?

RobertBColton commented on CompilerSource/ipc/Sprite.fbs:
```diff
@@ -0,0 +1,39 @@
+  shape:SpriteShape;
+  alpha_tolerance:int;
+  separate_mask:bool;
+  smooth_edges:bool;
```
Do we want to drop this field for both Sprites and Backgrounds? I always complained it never did anything in GM, and surprise surprise, it doesn't do anything in ENIGMA either. I don't think there's any graphical algorithm you can apply to 2D sprites to make them smoother, I don't think this flag is the same thing as interpolation. That itself is another problem because in older OpenGL's, interpolation is per-texture, whereas Direct3D has always been per-sampler (of which there are 8). Modern OpenGL has an extension to do it both ways.

RobertBColton commented on CompilerSource/ipc/Sprite.fbs:
```diff
@@ -0,0 +1,39 @@
+
+enum SpriteShape : byte { PRECISE, RECTANGLE, DISK, DIAMOND }
+
+enum SpriteBoundingBoxType : byte { AUTOMATIC, FULL_IMAGE, MANUAL }
```
Is SpriteBoundingBoxType a little long for yah? I prefer naming all of this "Type" over "Kind" because it's just less awkward. Kind is more often used to link something as belonging to a group, not to name the group itself. I don't know which is more semantically correct. Kind is also more informal than Type. Thus it seems it would be more appropriate to postfix "Kind" to the field name and "Type" to field type?

RobertBColton commented on CompilerSource/ipc/GameInformation.fbs:
```diff
@@ -0,0 +1,18 @@
+  left:int;
+  top:int;
+  width:int;
+  height:int;
```
Notice this uses left, top, width, and height, whereas the SpriteBoundingBox uses left, top, right, and bottom. Maybe we could reuse the same BoundingBox here and just have Sprite wrap around it to add its "Type"?

RobertBColton commented on CompilerSource/ipc/Background.fbs:
```diff
@@ -0,0 +1,22 @@
+  horizontal_offset:int;
+  vertical_offset:int;
+  horizontal_spacing:int;
+  vertical_spacing:int;
```
I could change this to the following now that Sprite has a Polygon subresource with a Point Type:
```flatbuffers
tile_size:Point;
offset:Point;
spacing:Point;
```
Should I do that? Would it make it clearer and more concise?

RobertBColton commented on CompilerSource/ipc/Room.fbs:
```diff
@@ -0,0 +1,113 @@
+  rotation:double;
+  color:int;
+
+  creation_code:string;
```
So there was a pre_creation_code before in our old backend that does not exist in GMS. I believe that was because it was easier for you and egofree to add color and rotation that way instead of updating the plugin. Since we have this now, I've just added the fields they would be since this mechanism is intended to be easier to update in the future. Or do you still find pre_creation_code kind of useful enough to include it too?

RobertBColton commented on CompilerSource/ipc/Room.fbs:
```diff
@@ -0,0 +1,113 @@
+  scale_x:double;
+  scale_y:double;
+  rotation:double;
+  color:int;
```
In the UI, the user selects both a color and an alpha for _image_blend_ and _image_alpha_ respectively. But for our buffers we can combine them into a single int, which GMX does. Is this right or should we make alpha a separate field?

RobertBColton commented on CompilerSource/ipc/Room.fbs:
```diff
@@ -0,0 +1,113 @@
+  id:int;
+
+  x:int;
+  y:int;
```
Again, I could be just using the Point struct here.

RobertBColton commented on CompilerSource/ipc/Room.fbs:
```diff
@@ -0,0 +1,113 @@
+  border_v:int;
+  speed_h:int;
+  speed_v:int;
+  object_id:int;
```
Another good place to reuse Point struct or the hypothetical BoundingBox struct I proposed to reuse from Sprite. Should I do this? It would make this schema a lot more concise.

RobertBColton commented on CompilerSource/ipc/Room.fbs:
```diff
@@ -0,0 +1,113 @@
+  editor_height:int;
+  current_tab:int;
+  scroll_x:int;
+  scroll_y:int;
```
Now, I know we said we want to keep the editor metadata to serialize it. But this is really only important for the other two groups of fields below this which are genuine settings and not plain old metadata.

The editor size itself is something we would honestly want the IDE to remember for **_EVERY_** editor. GMX does serialize this in the GMX. But I'm thinking of something bigger that would serialize the entire state of the IDE for every editor and dock widget when you close the window.

The disadvantage to this is then that it gets serialized to the registry or wherever a settings object in wx or Qt would put it thus making it less portable. The disadvantage is that if you take your project somewhere and open it on another computer, the entire layout is forgotten.

One possibility would be to continue serializing all of this in the project, but put it in a separate metadata file at the root of the project. The reason being is it's junk and **_people don't want that in their git logs_**. This is actually the case with a Qt or CodeBlocks project itself where I added the persistent layout to gitignore.

RobertBColton commented on CompilerSource/ipc/Room.fbs:
```diff
@@ -0,0 +1,113 @@
+
+  scale_x:double;
+  scale_y:double;
+  color:int;
```
For some reason they didn't give tiles rotation like they did instances. Maybe we should change that? Is it any less efficient when batching the tiles? I mean, I guess it really doesn't make sense to rotate tiles but.... it could be useful.

It seems Tiled has tile rotation:
http://discourse.mapeditor.org/t/can-i-rotate-tiles/703

So I don't see why not?

RobertBColton commented:
Closing because lol rip fatguy ur mom big gay I guess we are using protoc.
